Finjinon/PhotoCaptureViewController.swift

```diff
@@ -65,15 +70,78 @@ open class PhotoCaptureViewController: UIViewController, PhotoCollectionViewLayo
 
         view.backgroundColor = UIColor.black
 
-        previewView = UIView(frame: view.bounds)
+        NotificationCenter.default.addObserver(forName: NSNotification.Name.UIDeviceOrientationDidChange, object: nil, queue: nil) { (_) -> Void in
+            switch UIDevice.current.orientation {
+            case .faceDown, .faceUp, .unknown:
+                ()
+            case .landscapeLeft, .landscapeRight, .portrait, .portraitUpsideDown:
+                self.orientation = UIDevice.current.orientation
+                self.updateWidgetsToOrientation()
+            }
+        }
+    }
+
+    private func updateImagePickerButton() {
+        if imagePickerAdapter == nil {
+            if pickerButton != nil {
+                pickerButton?.removeFromSuperview()
+                pickerButton = nil
+            }
+        } else {
+            if pickerButton == nil {
+                let pickerButtonWidth: CGFloat = 114
+                pickerButton = UIButton(frame: CGRect(x: viewFrame.width - pickerButtonWidth - buttonMargin, y: viewFrame.origin.y + buttonMargin, width: pickerButtonWidth, height: 38))
+                pickerButton!.setTitle(NSLocalizedString("Photos", comment: "Select from Photos buttont itle"), for: UIControlState())
+                pickerButton!.setImage(UIImage(named: "PhotosIcon"), for: UIControlState())
+                pickerButton!.addTarget(self, action: #selector(presentImagePickerTapped(_:)), for: .touchUpInside)
+                pickerButton!.titleLabel?.font = UIFont.preferredFont(forTextStyle: UIFontTextStyle.footnote)
+                pickerButton!.autoresizingMask = [.flexibleTopMargin]
+                pickerButton!.layer.anchorPoint = CGPoint(x: 0.5, y: 0.5)
+                roundifyButton(pickerButton!)
+                view.addSubview(pickerButton!)
+            }
+        }
+    }
+
+    open override func viewWillAppear(_ animated: Bool) {
+        super.viewWillAppear(animated)
+
+        // In case the application uses the old style for managing status bar appearance
+        UIApplication.shared.setStatusBarHidden(true, with: .slide)
+    }
+
+    open override func viewDidAppear(_ animated: Bool) {
+        super.viewDidAppear(animated)
+        if #available(iOS 11.0, *) {
+            view.insetsLayoutMarginsFromSafeArea = true
+            viewFrame = view.convert(view.safeAreaLayoutGuide.layoutFrame, to: view.superview ?? view)
+            viewBounds = view.safeAreaLayoutGuide.layoutFrame
+        } else {
+            viewFrame = view.frame
+            viewBounds = view.bounds
+        }
+
+        setupSubviews()
+
+        collectionView.reloadData()
+        scrollToLastAddedAssetAnimated(false)
+    }
+
+    func setupSubviews() {
```

Comment:
Moving the setup of several implicitly unwrapped viewObjects _from_ `viewDidLoad` seems _very_ risky.  I think this needs to be reconsidered.

Reply:
Good point. I'll instantiate them directly during declaration of the class' properties,  so they'll never be nil; and then set frame after the views sizes and safe areas are known during viewDidAppear.